storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java

```diff
@@ -0,0 +1,671 @@
+     * transaction index, but the completion must be done only after successfully appending to the index.
+     */
+    public long lastStableOffset(CompletedTxn completedTxn) {
+        return findNextIncompleteTxn(completedTxn.producerId).map(x -> x.firstOffset.messageOffset).orElse(completedTxn.lastOffset + 1);
+    }
+
+    private Optional<TxnMetadata> findNextIncompleteTxn(long producerId) {
+        for (TxnMetadata txnMetadata : ongoingTxns.values()) {
+            if (txnMetadata.producerId != producerId) {
+                return Optional.of(txnMetadata);
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Mark a transaction as completed. We will still await advancement of the high watermark before
+     * advancing the first unstable offset.
+     */
+    public void completeTxn(CompletedTxn completedTxn) {
+        TxnMetadata txnMetadata = ongoingTxns.remove(completedTxn.firstOffset);
+        if (txnMetadata == null)
+            throw new IllegalArgumentException("Attempted to complete transaction $completedTxn on partition $topicPartition " + "which was not started");
+
+        txnMetadata.lastOffset = OptionalLong.of(completedTxn.lastOffset);
+        unreplicatedTxns.put(completedTxn.firstOffset, txnMetadata);
+        updateOldestTxnTimestamp();
+    }
+
+    public void deleteSnapshotsBefore(long offset) throws IOException {
+        for (SnapshotFile snapshot : snapshots.subMap(0L, offset).values()) {
+            removeAndDeleteSnapshot(snapshot.offset);
+        }
+    }
+
+    private Optional<SnapshotFile> oldestSnapshotFile() {
+        return Optional.ofNullable(snapshots.firstEntry()).map(x -> x.getValue());
+    }
+
+    private Optional<SnapshotFile> latestSnapshotFile() {
+        return Optional.ofNullable(snapshots.lastEntry()).map(e -> e.getValue());
+    }
+
+    /**
+     * Removes the producer state snapshot file metadata corresponding to the provided offset if it exists from this
+     * ProducerStateManager, and deletes the backing snapshot file.
+     */
+    private void removeAndDeleteSnapshot(long snapshotOffset) throws IOException {
+        SnapshotFile snapshotFile = snapshots.remove(snapshotOffset);
+        if (snapshotFile != null) snapshotFile.deleteIfExists();
+    }
+
+    /**
+     * Removes the producer state snapshot file metadata corresponding to the provided offset if it exists from this
+     * ProducerStateManager, and renames the backing snapshot file to have the Log.DeletionSuffix.
+     * <p>
+     * Note: This method is safe to use with async deletes. If a race occurs and the snapshot file
+     * is deleted without this ProducerStateManager instance knowing, the resulting exception on
+     * SnapshotFile rename will be ignored and None will be returned.
+     */
+    public Optional<SnapshotFile> removeAndMarkSnapshotForDeletion(long snapshotOffset) throws IOException {
```

Comment:
Did we make this public due to the callers being in different files (previously same package, but no longer)?

Reply:
Right.